core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -300,6 +310,8 @@ pub(super) enum Misbehavior {
 	// A message received that's from the future relative to our view.
 	// always misbehavior.
 	FutureMessage,
+    // Equivocation received
+    Equivocation,
```

Comment:
(more spaces, here and elsewhere)

Reply:
also, sending an equivocation is _not_ misbehavior (sometimes, equivocations are required to complete a round). sending too many equivocations is misbehavior

Reply:
Should I perhaps instead do something using the `_equivocation` "callbacks" on `Environment`, as opposed to determining that something is an equivocation just by counting incoming vote messages? 

It appears to me that the only way to know for sure if a message is indeed an equivocation is by using that handle to get the feedback from the actual grandpa worker.

Also, how do we determine if something is "too many" equivocations?

Reply:
Since a single equivocation is counted as all possible equivocations, we want to ignore all except the first because they are redundant.

We know something is an equivocation even without the callbacks because the definition is 2 messages by the same voter of the same type in the same round.